src/echoclient/EchoClient.php

```diff
@@ -62,51 +65,45 @@ function __construct()
     }
 
     /**
-     * Add an event to echo
+     * Adds a single event to echo
      *
      * @param string $class
      * @param string $source
      * @param array $props
      * @param string|null $userId
+     * @param string|null $timestamp
      * @return bool True if successful, else false
      */
-    public function createEvent($class, $source, array $props = array(), $userId = null)
+    public function createEvent($class, $source, array $props = array(), $userId = null, $timestamp = null)
     {
         $class = ECHO_CLASS_PREFIX . $class;
-        $baseUrl = $this->getBaseUrl();
+        $eventJson = $this->getEventJson($class, $source, $props, $userId, $timestamp);
 
-        if (!$baseUrl) {
-            // fail silently when creating events, should not stop user interaction as echo events are collected on a best-endeavours basis
-            $this->getLogger()->warning('Echo server is not defined (missing ECHO_HOST define), not sending event - ' . $class,
-                $props);
-            return false;
-        }
-
-        $eventUrl = $baseUrl . '/events';
-        $eventJson = $this->getEventJson($class, $source, $props, $userId);
+        return $this->sendJsonEventDataToEcho($eventJson);
+    }
 
-        try {
-            $client = $this->getHttpClient();
-            $request = $client->post($eventUrl, $this->getHeaders(), $eventJson, array('connect_timeout' => 2));
-            $response = $request->send();
+    /**
+     * Adds multiple events to echo
+     *
+     * @param EchoEvent[] $events An array of EchoEvent objects
+     * @return bool True if successful, else false
+     */
+    public function createEvents(array $events)
+    {
+        if (count($events) > self::ECHO_MAX_BATCH_EVENTS) {
+            $this->getLogger()->warning('Batch contains more than ' . self::ECHO_MAX_BATCH_EVENTS . ' events');
```

Comment:
do we need this warning? It will end up filling the logs with information that we most likely don't care about.

Reply:
I only added it to be consistent with other messages, but I agree we don't need it, but by that token we should look at some of the others as well.

Reply:
If the app start's erroring because of this exception - we probably want the warning (even error?) so we know what's causing the app to fail don't we?

Reply:
In my opinion, this should be an exception.  The events are in my hands as the user of this client, and I've passed them to this method, and the method should tell me I've done a bad thing.  Without an exception, I cannot tell why my events were not sent.

I understand that we may want to silence the exceptions one-by-one in the old method, but I'd advocate we reconsider this for the new methods.  It seems crazy that the feedback here is effectively

http://gifimage.net/wp-content/uploads/2017/10/computer-says-no-gif-2.gif

Reply:
Oh look at that - one line below.  A new personal record in dumbassery.

Reply:
I totally agree it should be an exception - which it is below. 

Is what happens in practice though that Exceptions when posting events to echo are swallowed because we don't want the failure to trigger an error in the app. For that reason, I think we should definitely have the log line.

Reply:
Isn't that the calling code's choice, though?

Reply:
Malc and I discussed this over a tea, and (I think?) agreed that:

* We should make the new methods present exceptions where the calling code has done something wrong;

* We should use some short logging to aid Platform team in tracing through issues (the client is part-proof of the calling code sending good events - silent fail in the client code would otherwise give us nothing to go on).

I hope that seems reasonable.

Reply:
@malc @junglebarry that's done now